### DISCUSSION TITLE
Deep reactivity in pyscript

### DISCUSSION THREAD

**Discussion by Neon22** [micropython/micropython#15711] [Webassembly]
Some people are trying to implement deep reactivity in pyscript. This is so we can have a reactive UI layer in the browser (similar to say Vue3). They want the pyodide and micropython interpreters to work the same for obvious reasons. The pyodide side is working but the micropython side is having some problems.

The discussion on achieving this is happening primarily in this thread:
- https://github.com/kkinder/puepy/discussions/17

There seems to be some confusion about what is supported with conflicting references and support for/about such things as:
 - `__get_attribute__`, `__setitem__`, `__setattr__`

There is functioning code (interactive, clonable) here: 
- https://pyscript.com/@mikaelho/reactive-data-test/latest?files=main.py,reactive_data.py

E.g. @ntoll has said:
"MicroPython does support those dunder methods (I've used them). Not sure what's going on. Perhaps ping MP folk with an example of the problem..?"

So I am posting here in the hopes someone can join our discussion and resolve our problem.
Cheers...

**Comment by stinos** [comment]
Some of those might not be implemented at all, others conditionally depending on compilation flags. Which can be different depending on which port/MCU is being used. So it's possible someone reports something as working whereas someone with another MCU doesn't have the feature by default but needs to enable it by building and passing flags like `CFLAGS_EXTRA=-DMICROPY_PY_DELATTR_SETATTR=1`.

As such there's not a single go-to place in the docs or so where one can find out if a feature is supported.
I usually grep the source code to figure that out. Or you could start with checking https://github.com/micropython/micropython/blob/master/py/mpconfig.h by searching for the feature name, then comparing with a certain port's mpconfigport.h, but I'm not sure everything there is listed by exact function name. Note for features which are implemented there can be differences with CPython: http://docs.micropython.org/en/latest/genrst/index.html.

**Reply by WebReflection** [reply]
I believe here everyone is talking about the WebAssembly for Pyscript port, which AFAIK is mostly maintained by @dpgeorge . Thanks for the rest of the details though, interesting regardless 🙏

**Reply by stinos** [reply]
> WebAssembly for Pyscript port

This should have pretty much all features because it's mpconfigvariant.h (sorry forgot to mention that next to mpconfigport.h there's also variants..) defines `#define MICROPY_CONFIG_ROM_LEVEL                (MICROPY_CONFIG_ROM_LEVEL_FULL_FEATURES)` 

But 'all features' isn't the same as CPython-compatible. If you guys need something implemented or want clarification it's best to post MCVE in this discussion or as an issue (if it doesn't exist yet)

**Comment by mikaelho** [comment]
Hello. To clarify a little bit what we trying to achieve with reactivity, and what our current main challenge is:

We have taken some CPython proxying code and use it to wrap the "watched" data structure in order to detect all mutations, which then trigger UI changes.

We have been able to work around the getattr differences. It would be nice if we could also enable `isinstance` in a way where it would match the actual object and not the proxy. This has proved difficult as MicroPython seems to skip `.__class__` and assumes that it is the same as calling `type()` (based on limited understanding of [this code](https://github.com/micropython/micropython/blob/6c3dc0c0b09701b83a957368220d9c8bfe40af0a/py/runtime.c#L1191)).

Clarifications for any misunderstanding on my part or hints on how to work around this difference would be much appreciated. Even better, of course, if `__class__` could become part of the `isinstance` check.

**Comment by laffra** [comment]
In case this helps, PySheets (which uses PyScript with MicroPython for the UI) uses a Model-View architecture for the cells in a spreadsheet. I override __setattr__ to inform listeners that something changed in the model. This works fine on "standard" PyScript releases for me. See https://github.com/PySheets/pysheets/blob/main/src/static/models.py#L184
